components/layout/display_list/builder.rs

```diff
@@ -1319,110 +1330,102 @@ impl FragmentDisplayListBuilding for Fragment {
-                        style,
-                        bounds.size,
-                        &gradient.items[..],
-                        shape,
-                        center,
-                        gradient.repeating,
-                    );
-                    stops = radial_stops;
-                    BorderDetails::NinePatch(webrender_api::NinePatchBorder {
-                        source: webrender_api::NinePatchBorderSource::RadialGradient(wr_gradient),
-                        width: 0,
-                        height: 0,
-                        slice: SideOffsets2D::zero(),
-                        fill: false,
-                        repeat_horizontal: webrender_api::RepeatMode::Stretch,
-                        repeat_vertical: webrender_api::RepeatMode::Stretch,
-                        outset: outset_layout,
-                    })
-                },
-            }),
-            Either::Second(Image::PaintWorklet(ref paint_worklet)) => self
-                .get_webrender_image_for_paint_worklet(state, style, paint_worklet, size)
-                .and_then(|image| {
-                    build_image_border_details(image, border_style_struct, outset_layout)
-                }),
-            Either::Second(Image::Rect(..)) => {
-                // TODO: Handle border-image with `-moz-image-rect`.
-                None
+            Either::First(_) => {
+                if border_widths == SideOffsets2D::zero() {
+                    None
+                } else {
+                    Some(BorderDetails::Normal(NormalBorder {
+                        left: BorderSide {
+                            color: style.resolve_color(colors.left).to_layout(),
+                            style: border_style.left.to_layout(),
+                        },
+                        right: BorderSide {
+                            color: style.resolve_color(colors.right).to_layout(),
+                            style: border_style.right.to_layout(),
+                        },
+                        top: BorderSide {
+                            color: style.resolve_color(colors.top).to_layout(),
+                            style: border_style.top.to_layout(),
+                        },
+                        bottom: BorderSide {
+                            color: style.resolve_color(colors.bottom).to_layout(),
+                            style: border_style.bottom.to_layout(),
+                        },
+                        radius: border_radius,
+                    }))
+                }
             },
-            Either::Second(Image::Element(..)) => {
-                // TODO: Handle border-image with `-moz-element`.
-                None
+            Either::Second(ref image) => {
+                build_border_image_details(
+                    bounds,
+                    border_widths,
+                    border_style_struct,
+                    |fallback_size| match image {
```

Comment:
It's a bit awkward to have this closure when we only call the function once... Maybe it'd be clearer to move `build_border_image_details` to operate on `self`, or take the required members as arguments?